### PR TITLE
Adding transformer method to PipelineModel

### PR DESCRIPTION
**Pull request by andremoeller**
*Issue #, if available:*

https://github.com/aws/sagemaker-python-sdk/issues/661

*Description of changes:*

Adds `transformer` to PipelineModel, allowing you to run multi-container Batch Transform jobs from PipelineModel (without having already created a model).

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [X] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [X] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [X] I have updated the [changelog](https://github.com/aws/sagemaker-python-sdk/blob/master/CHANGELOG.rst) with a description of my changes (if appropriate)
- [X] I have updated any necessary [documentation](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/pipeline.py [hunk 1]
```diff
@@ -49,6 +50,14 @@ def __init__(self, models, role, predictor_cls=None, name=None, vpc_config=None,
         self.sagemaker_session = sagemaker_session
         self._model_name = None
 
+    def enable_network_isolation(self):
+        """Whether to enable network isolation when creating this Model
+
+        Returns:
+            bool: If network isolation should be enabled or not.
+        """
+        return False
```
Comment (andremoeller) [inline src/sagemaker/pipeline.py]: I copied this from model.py but didn't really understand it -- is this just supposed to always be false?

Comment (iquintero) [reply src/sagemaker/pipeline.py]: For now yes it should be false. Only Models that are consumed from Marketplace set this to True.

Comment (andremoeller) [reply src/sagemaker/pipeline.py]: Ah, ok

File: src/sagemaker/model.py [hunk 2]
```diff
@@ -285,10 +285,6 @@ def transformer(self, instance_count, instance_type, strategy=None, assemble_wit
             max_payload (int): Maximum size of the payload in a single HTTP request to the container in MB.
             tags (list[dict]): List of tags for labeling a transform job. If none specified, then the tags used for
                 the training job are used for the transform job.
-            role (str): The ``ExecutionRoleArn`` IAM Role ARN for the ``Model``, which is also used during
```
Comment (orchidmajumder) [inline src/sagemaker/model.py]: I am not being able to understand the connection of removal of these lines to this PR. Are these changes intentional?

File: src/sagemaker/pipeline.py [hunk 3]
```diff
@@ -104,6 +113,56 @@ def deploy(self, initial_instance_count, instance_type, endpoint_name=None, tags
         if self.predictor_cls:
             return self.predictor_cls(self.endpoint_name, self.sagemaker_session)
 
+    def _create_sagemaker_pipeline_model(self, instance_type):
+        """Create a SageMaker Model Entity
+
+        Args:
+            instance_type (str): The EC2 instance type that this Model will be used for, this is only
+                used to determine if the image needs GPU support or not.
+            accelerator_type (str): <put docs here>
```
Comment (orchidmajumder) [inline src/sagemaker/pipeline.py]: nit: Please remove this before this variable is added to the constructor.

File: src/sagemaker/pipeline.py [hunk 4]
```diff
@@ -104,6 +113,56 @@ def deploy(self, initial_instance_count, instance_type, endpoint_name=None, tags
...
+                each individual transform container at one time.
+            max_payload (int): Maximum size of the payload in a single HTTP request to the container in MB.
+            tags (list[dict]): List of tags for labeling a transform job. If none specified, then the tags used for
+                the training job are used for the transform job.
+            volume_kms_key (str): Optional. KMS key ID for encrypting the volume attached to the ML
+                compute instance (default: None).
+        """
+        self._create_sagemaker_pipeline_model(instance_type)
+        if self.enable_network_isolation():
+            env = None
```
Comment (orchidmajumder) [inline src/sagemaker/pipeline.py]: Is this a requirement for internet-free mode? Multi container setup depends on environment variable to retrieve the `Accept` types for the containers (apart from the last one, which receives the original `Accept` from request).

File: tests/unit/test_pipeline_model.py [hunk 5]
```diff
@@ -111,17 +111,58 @@ def test_deploy_endpoint_name(tfo, time, sagemaker_session):
     framework_model = DummyFrameworkModel(sagemaker_session)
     sparkml_model = SparkMLModel(model_data=MODEL_DATA_2, role=ROLE, sagemaker_session=sagemaker_session)
     model = PipelineModel(models=[framework_model, sparkml_model], role=ROLE, sagemaker_session=sagemaker_session)
-    model.deploy(endpoint_name='blah', instance_type=INSTANCE_TYPE, initial_instance_count=55)
+    model.deploy(instance_type=INSTANCE_TYPE, initial_instance_count=1)
     sagemaker_session.endpoint_from_production_variants.assert_called_with(
-        'blah',
+        'mi-1-2017-10-10-14-14-15',
```
Comment (orchidmajumder) [inline tests/unit/test_pipeline_model.py]: Where is this string coming from?

File: src/sagemaker/model.py (hunk 2 above)
Comment (andremoeller) [reply src/sagemaker/model.py]: Yes. There's no `role` or `model_server_workers` parameters in this method. (The docstring was probably just copied from the estimator `transformer` method, which does have these parameters)

File: src/sagemaker/pipeline.py (hunk 3 above)
Comment (andremoeller) [reply src/sagemaker/pipeline.py]: Oh dear. Will do.

File: tests/unit/test_pipeline_model.py (hunk 5 above)
Comment (andremoeller) [reply tests/unit/test_pipeline_model.py]: Seems like it's 'model image 1' followed by a timestamp.

That's not actually part of my diff, though. I've only added test_transformer -- git's getting it wrong since the two methods start the same way.

File: src/sagemaker/pipeline.py (hunk 4 above)
Comment (andremoeller) [reply src/sagemaker/pipeline.py]: I'm not sure. I'll see if I can remove this.

Comment (andremoeller) [reply src/sagemaker/pipeline.py]: Either way, I talked it over with @iquintero and it's better to remove it.

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://9dzx6wxcy9.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=375d7344-d55b-4fcd-8b55-8a455ae0c619%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by andremoeller** [comment]
I addressed @orchidmajumder s comments -- could an approved reviewer help move this PR forward?

Thanks!

**Comment by orchidmajumder** [comment]
@andremoeller Github is telling me that there is no new changes since my last review. Have you updated your remote tracking branch to take care of the comments?

**Comment by andremoeller** [comment]
@orchidmajumder Yes. I've force pushed (bad habit -- just trying to keep to one commit), so GitHub probably doesn't have the previous commits.

**Comment by andremoeller** [comment]
Ping -- is there anything else to do before this can be merged in? Thanks.

File: tests/integ/test_inference_pipeline.py [hunk 6]
```diff
@@ -87,11 +123,13 @@ def test_inference_pipeline_model_deploy(sagemaker_session):
                                       serializer=json_serializer, content_type=CONTENT_TYPE_CSV,
                                       accept=CONTENT_TYPE_CSV)
 
-        valid_data = '1.0,C,38.0,71.5,1.0,female'
-        assert predictor.predict(valid_data) == "0.714013934135"
+        with open(VALID_DATA_PATH, 'r') as f:
+            valid_data = f.read()
+            assert predictor.predict(valid_data) == "0.714013934135"
```
Comment (laurenyu) [inline tests/integ/test_inference_pipeline.py]: nit: use single quotes

File: tests/data/sparkml_xgboost_pipeline/invalid_input.csv [hunk 7]
```diff
@@ -0,0 +1 @@
+1.0,28.0,C,38.0,71.5,1.0
```
Comment (laurenyu) [inline tests/data/sparkml_xgboost_pipeline/invalid_input.csv]: nit: add a newline to the end of each of these CSVs

File: CHANGELOG.rst [hunk 8]
```diff
@@ -8,11 +8,12 @@ CHANGELOG
 
 * doc-fix: Remove incorrect parameter for EI TFS Python README
 * feature: ``Predictor``: delete SageMaker model
-* feature: ``Pipeline``: delete SageMaker model
+* feature: ``PipelineModel``: delete SageMaker model
 * bug-fix: Estimator.attach works with training jobs without hyperparameters
 * doc-fix: remove duplicate content from mxnet/README.rst
 * doc-fix: move overview content in main README into sphynx project
 * bug-fix: pass accelerator_type in ``deploy`` for REST API TFS ``Model``
+* feature: ``PipelineModel``: Create a Transformer from a PipelineModel
```
Comment (laurenyu) [inline CHANGELOG.rst]: 1.18.4 has been released, so this will need to go under a new version

**Comment by andremoeller** [comment]
OK, fixed.

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://9dzx6wxcy9.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=b3727156-f66e-44f5-a615-0919fbc592ed%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://9dzx6wxcy9.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=db2a9146-7b3d-44ca-8a5e-e39c971d5d63%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*
